### PR TITLE
NDRS-706: Provide JSON representations for Contract and ContractPackage

### PR DESCRIPTION
**Pull request by darthsiroftardis**
Created a new PR from: https://github.com/CasperLabs/casper-node/pull/718. 

There was some back and forth on certain representations for json_compatibility. This PR contains a cleaner commit stream and the optimal implementation.

REF: https://casperlabs.atlassian.net/browse/NDRS-706

Added `Contract` and `ContractPackage` struct to a new contracts.rs module inside the json compatibility module.

### CODE REVIEW

File: types/src/key.rs [hunk 1]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
 }
 
 /// An alias for [`Key`]s hash variant.
-pub type ContractHash = HashAddr;
+pub type ContractHashBytes = HashAddr;
```
Comment (Fraser999) [inline types/src/key.rs:70]: Does this need to be public?  I don't think we even need this alias tbh.  I'd just use `HashAddr`.  Same for `ContractPackageHashBytes` and `ContractWasmHashBytes`.

File: types/src/key.rs [hunk 2]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
-pub type ContractPackageHash = HashAddr;
+pub type ContractPackageHashBytes = HashAddr;
+
+/// A newtype wrapping a `ContractPackageHashBytes` which is the raw bytes of
+/// the ContractPackageHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractPackageHash(ContractPackageHashBytes);
+
+impl ContractPackageHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
```
Comment (Fraser999) [inline types/src/key.rs:374]:
```suggestion
    /// Constructs a new `ContractPackageHash` from the raw bytes of the contract hash.
```

Also, as above, please change "contract hash" to "contract package hash" throughout comments.

File: types/src/key.rs [hunk 3]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    pub const fn new(value: ContractPackageHashBytes) -> ContractPackageHash {
+        ContractPackageHash(value)
+    }
+    /// Returns the raw bytes of the contract hash as an array.
```
Comment (Fraser999) [inline types/src/key.rs:378]: Blank lines.

File: types/src/key.rs [hunk 4]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    }
+    /// Returns the raw bytes of the contract hash as a `slice`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+    /// Formats the `ContractHash` for users getting and putting.
+    pub fn to_formatted_string(&self) -> String {
+        format!("{}{}", PACKAGE_STRING_PREFIX, base16::encode_lower(&self.0),)
+    }
+    /// Parses Parses a string formatted as per `Self::to_formatted_string()` into an
```
Comment (Fraser999) [inline types/src/key.rs:390]:
```suggestion
    /// Parses a string formatted as per `Self::to_formatted_string()` into a
```

File: types/src/key.rs [hunk 5]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    type Error = TryFromSliceForContractHashError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, TryFromSliceForContractHashError> {
+        ContractWasmHashBytes::try_from(bytes)
+            .map(ContractWasmHash::new)
+            .map_err(|_| TryFromSliceForContractHashError(()))
+    }
+}
+
+impl TryFrom<&alloc::vec::Vec<u8>> for ContractWasmHash {
```
Comment (Fraser999) [inline types/src/key.rs:340]:
```suggestion
impl TryFrom<&Vec<u8>> for ContractWasmHash {
```

File: types/src/key.rs [hunk 6]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    type Error = TryFromSliceForContractHashError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, TryFromSliceForContractHashError> {
+        ContractPackageHashBytes::try_from(bytes)
+            .map(ContractPackageHash::new)
+            .map_err(|_| TryFromSliceForContractHashError(()))
+    }
+}
+
+impl TryFrom<&alloc::vec::Vec<u8>> for ContractPackageHash {
```
Comment (Fraser999) [inline types/src/key.rs:482]:
```suggestion
impl TryFrom<&Vec<u8>> for ContractPackageHash {
```

**Review by Fraser999 — COMMENTED** [review]
The types `ContractHash` and `ContractPackageHash` really belong in the `contract` module, and `ContractPackageHash` belongs in the `contract_wasm` module.  Sorry I didn't notice that sooner.

If you feel that the contract module should be split over several submodules in separate files, feel free to make that change too.

File: node/src/types/json_compatibility/stored_value.rs [hunk 7]
```diff
@@ -16,14 +16,15 @@ use casper_types::{
...
 };
 
-use super::Account;
+use super::{Account, Contract, ContractPackage};
 
 /// Representation of a value stored in global state.
 ///
-/// All variants other than `Account` are formed by serializing the variant using `ToBytes` and hex-
-/// encoding the resulting byte string.
+/// Variants like `Account`, `Contract` and `ContractPackage` are formed by serializing the variant
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/stored_value.rs:23]: No, it applies to `Contract` and `ContractPackage`

File: node/src/types/json_compatibility/stored_value.rs [hunk 8]
```diff
@@ -16,14 +16,15 @@ use casper_types::{
...
+/// using `ToBytes` and hex- encoding the resulting byte string.
 ///
-/// `Account` has its own `json_compatibility` representation (see its docs for further info).
+/// `Account`, `Contract` and `ContractPackage` have their own `json_compatibility` representation
+/// (see its docs for further info).
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/stored_value.rs]: Applied suggestion

File: types/src/key.rs [hunk 9]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+/// the ContractHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractHash(ContractHashBytes);
+
+impl ContractHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
+    pub const fn new(value: ContractHashBytes) -> ContractHash {
+        ContractHash(value)
+    }
+    /// Returns the raw bytes of the contract hash as an array.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Blank lines have been inserted.

File: types/src/key.rs [hunk 10]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    }
+    /// Formats the `ContractHash` for users getting and putting.
+    pub fn to_formatted_string(&self) -> String {
+        format!(
+            "{}{}",
+            CONTRACT_STRING_PREFIX,
+            base16::encode_lower(&self.0),
+        )
+    }
+    /// Parses Parses a string formatted as per `Self::to_formatted_string()` into an
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: types/src/key.rs [hunk 11]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    type Error = TryFromSliceForContractHashError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, TryFromSliceForContractHashError> {
+        ContractHashBytes::try_from(bytes)
+            .map(ContractHash::new)
+            .map_err(|_| TryFromSliceForContractHashError(()))
+    }
+}
+
+impl TryFrom<&alloc::vec::Vec<u8>> for ContractHash {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Changed level of qualification

File: types/src/key.rs [hunk 12]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+/// the ContractWasmHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractWasmHash(ContractWasmHashBytes);
+
+impl ContractWasmHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
+    pub const fn new(value: ContractWasmHashBytes) -> ContractWasmHash {
+        ContractWasmHash(value)
+    }
+    /// Returns the raw bytes of the contract hash as an array.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Blank lines added.

File: types/src/key.rs [hunk 13]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
-pub type ContractWasmHash = HashAddr;
+pub type ContractWasmHashBytes = HashAddr;
+
+/// A newtype wrapping a `ContractWasmHashBytes` which is the raw bytes of
+/// the ContractWasmHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractWasmHash(ContractWasmHashBytes);
+
+impl ContractWasmHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Changed comments

File: types/src/key.rs [hunk 14]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    }
+    /// Returns the raw bytes of the contract hash as an array.
+    pub fn value(&self) -> ContractWasmHashBytes {
+        self.0
+    }
+    /// Returns the raw bytes of the contract hash as a `slice`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+    /// Formats the `ContractHash` for users getting and putting.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: types/src/key.rs [hunk 15]
```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
...
+    pub fn to_formatted_string(&self) -> String {
+        format!("{}{}", WASM_STRING_PREFIX, base16::encode_lower(&self.0),)
+    }
+    /// Parses Parses a string formatted as per `Self::to_formatted_string()` into an
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: types/src/key.rs (hunk 1 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Removed aliases and using `HashAddr` directly

File: types/src/key.rs (hunk 2 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Changed comments

File: types/src/key.rs (hunk 3 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Added blank lines

File: types/src/key.rs (hunk 4 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: types/src/key.rs (hunk 5 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: types/src/key.rs (hunk 6 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Applied suggestion

File: node/src/types/json_compatibility/stored_value.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/types/json_compatibility/stored_value.rs:23]: I don't agree.  We used to just serialize these types to a byte array and hex-encode, but that's not what we do now.  We convert it to a JSON-compatible type.

File: node/src/types/json_compatibility/contracts.rs [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+use datasize::DataSize;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::types::json_compatibility::vectorize;
+use casper_types::{
+    Contract as DomainContract, ContractHash, ContractPackage as DomainContractPackage,
+    ContractPackageHash, ContractWasmHash, EntryPoint, NamedKey, URef,
+};
+use semver::Version;
```
Comment (Fraser999) [inline node/src/types/json_compatibility/contracts.rs:13]: This should be with the other 3rd party imports.

File: execution_engine/src/core/engine_state/system_contract_cache.rs [hunk 17]
```diff
@@ -3,9 +3,8 @@ use std::{
     sync::{Arc, RwLock},
 };
 
-use parity_wasm::elements::Module;
-
 use casper_types::ContractHash;
+use parity_wasm::elements::Module;
```
Comment (Fraser999) [inline execution_engine/src/core/engine_state/system_contract_cache.rs:7]: This was correct before - parity_wasm is a 3rd party crate.

File: types/src/key.rs [hunk 18]
```diff
@@ -57,12 +59,15 @@ impl From<HashAddr> for Key {
...
-pub type ContractPackageHash = HashAddr;
+/// Associated error type of `TryFrom<&[u8]>` for `ContractHash`.
+#[derive(Debug)]
+pub struct TryFromSliceForContractHashError(());
+
+impl Display for TryFromSliceForContractHashError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "failed to retrieve from slice")
+    }
+}
```
Comment (Fraser999) [inline types/src/key.rs:70]: This should be deleted I think?

File: types/src/contracts.rs [hunk 19]
```diff
@@ -170,6 +259,295 @@ pub type DisabledVersions = BTreeSet<ContractVersionKey>;
...
+    }
+}
+
+/// A newtype wrapping a `HashAddr` which is the raw bytes of
+/// the ContractPackageHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractPackageHash(HashAddr);
+
+impl ContractPackageHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
```
Comment (Fraser999) [inline types/src/contracts.rs:414]:
```suggestion
    /// Constructs a new `ContractPackageHash` from the raw bytes of the contract hash.
```

You forgot to refactor all "contract hash" comments to "contract package hash".

File: types/src/contract_wasm.rs [hunk 20]
```diff
@@ -1,11 +1,226 @@
...
+    }
+}
+
+/// A newtype wrapping a `HashAddr` which is the raw bytes of
+/// the ContractWasmHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractWasmHash(HashAddr);
+
+impl ContractWasmHash {
+    /// Constructs a new `ContractWasmHash` from the raw bytes of the contract hash.
```
Comment (Fraser999) [inline types/src/contract_wasm.rs:89]: You forgot to replace "contract hash" with "contract wasm hash" in the comments.

File: node/src/types/json_compatibility/contracts.rs (hunk 16 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/contracts.rs]: Regrouped

File: execution_engine/src/core/engine_state/system_contract_cache.rs (hunk 17 above)
Comment (darthsiroftardis) [reply execution_engine/src/core/engine_state/system_contract_cache.rs]: Regrouped

File: types/src/key.rs (hunk 18 above)
Comment (darthsiroftardis) [reply types/src/key.rs]: Deleted

File: types/src/contracts.rs (hunk 19 above)
Comment (darthsiroftardis) [reply types/src/contracts.rs]: Applied suggestion

File: types/src/contract_wasm.rs (hunk 20 above)
Comment (darthsiroftardis) [reply types/src/contract_wasm.rs]: Replaced

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**